### PR TITLE
Prevent suspicious directory traversal

### PR DESCRIPTION
**Pull request by fbricon**
Signed-off-by: Fred Bricon <fbricon@gmail.com>

### CODE REVIEW

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/uriresolver/CacheResourcesManager.java [hunk 1]
```diff
@@ -51,7 +52,7 @@
 public class CacheResourcesManager {
 
 	private static final String USER_AGENT_KEY = "User-Agent";
-	private static final String USER_AGENT_VALUE = "LemMinX";
+	private static final String USER_AGENT_VALUE = "LemMinX/"+Platform.getVersion().getVersionNumber() + " ("+Platform.getOS().getName() + " "+ Platform.getOS().getVersion()+")";
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/uriresolver/CacheResourcesManager.java:55]: I fear that we can have the same problem with Java/1.8 user agent which contains à.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/uriresolver/InvalidURIException.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * Invalid URI exception.
+ *
+ * @author Fred Bricon
+ *
+ */
+public class InvalidURIException extends CacheResourceException {
+
+	private static final long serialVersionUID = 1L;
+	
+	public InvalidURIException(String resourceURI, String message, Throwable cause) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/uriresolver/InvalidURIException.java:25]: To be consistent with cacheresourcedownloadingewception the message should be in this class

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/uriresolver/CacheResourcesManager.java (hunk 1 above)
Comment (fbricon) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/uriresolver/CacheResourcesManager.java:55]: No it's fine. I just tried: 

```java
	public static void main(String[] args) throws Exception {
		String actualURI = "https://lime.software/xsd/project-1.0.0.xsd";
		URL url = new URL(actualURI);
		URLConnection conn = url.openConnection();
		//conn.addRequestProperty("User-Agent", "Java/1.8"); // throws a 403
		conn.addRequestProperty("User-Agent", "LemMinX/1.8"); // works fine
		conn.getInputStream();
	}
```

Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/uriresolver/CacheResourcesManager.java:55]: What? How it is possible? Thanks for the info.

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/uriresolver/CacheResourcesManagerTest.java [hunk 3]
```diff
@@ -116,21 +121,46 @@ public void testAvailableCache() throws Exception {
...
-		CacheResourceDownloadingException actual = null;
+	public void testGetBadResourceName() throws Exception {
+		String url = "http://localhost/foo/bar/`test.txt`";
 		try {
 			cacheResourcesManager.getResource(url);
-		} catch (CacheResourceDownloadingException e) {
-			actual = e;
+			fail("Invalid url should fail to download");
+		} catch (Exception e) {
+			assertEquals(InvalidURIException.class, e.getClass());
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/uriresolver/CacheResourcesManagerTest.java:130]: Add an assert to check the getErrorCode

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/uriresolver/CacheResourcesManagerTest.java [hunk 4]
```diff
@@ -116,21 +121,46 @@ public void testAvailableCache() throws Exception {
...
+		assertEquals("choice.xsd",path.getFileName().toString());
+		String choice = FilesUtils.readString(path);
+		assertTrue(choice.contains("<xs:element name=\"person\">"), () -> {return "Unexpected file content:"+choice;});
+		
+		String invalidUri = server.getUri("/../../../xsd/choice.xsd");
+		try {
+			cacheResourcesManager.getResource(invalidUri);
+			fail("Invalid url should fail to download");
+		} catch (Exception e) {
+			assertEquals(InvalidURIException.class, e.getClass());
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/uriresolver/CacheResourcesManagerTest.java:155]: Add an assert to check the getErrorCode

**Comment by angelozerr** [comment]
Thanks @fbricon !

**Pull request merged**